cmd/buildctl/spawner/spawner_linux.go

```diff
@@ -0,0 +1,182 @@
+package spawner
+
+import (
+	"context"
+	"net"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+
+	sddaemon "github.com/coreos/go-systemd/daemon"
+	"github.com/google/shlex"
+	"github.com/moby/buildkit/util/appdefaults"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/errgroup"
+)
+
+type spawner struct {
+	flags           []string
+	buildkitdSocket string
+	closer          func() error
+}
+
+func New(flagsStr string) (Spawner, error) {
+	flags, err := shlex.Split(flagsStr)
+	if err != nil {
+		return nil, err
+	}
+	buildkitdAddr := appdefaults.Address
+	if os.Geteuid() != 0 {
+		buildkitdAddr = appdefaults.UserAddress()
+	}
+	if !strings.HasPrefix(buildkitdAddr, "unix://") {
+		return nil, errors.Errorf("unexpected non-unix address %q", buildkitdAddr)
+	}
+	buildkitdSocket := strings.TrimPrefix(buildkitdAddr, "unix://")
+	sp := &spawner{
+		flags:           flags,
+		buildkitdSocket: buildkitdSocket,
+	}
+	return sp, nil
+}
+
+func (sp *spawner) cmd(buildkitdAddr, notifySocket string) *exec.Cmd {
+	const (
+		buildkitd   = "buildkitd"
+		rootlesskit = "rootlesskit"
+	)
+	flags := append(sp.flags, "--addr="+buildkitdAddr)
+	var cmd *exec.Cmd
+	if os.Geteuid() == 0 {
+		cmd = exec.Command(buildkitd, flags...)
+	} else {
+		cmd = exec.Command(rootlesskit, append([]string{buildkitd}, flags...)...)
+	}
+	cmd.Env = append(os.Environ(), "NOTIFY_SOCKET="+notifySocket)
+	cmd.Stdout = &logrusWriter{debugPrefix: "spawn(stdout): "}
+	cmd.Stderr = &logrusWriter{debugPrefix: "spawn(stderr): "}
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid:   true,
```

Comment:
Setpgid so that we can kill the daemon along with all children processes.
The issue was actually resolved in https://github.com/moby/buildkit/pull/1008 but should not hurt to have setpgid as well